In [1]:
import pandas as pd
import numpy as np

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

from imblearn.over_sampling import SMOTE

In [3]:
import sklearn
print(sklearn.__version__)

1.7.1


In [4]:
import imblearn
print(imblearn.__version__)

0.14.0


In [5]:
df = pd.read_csv("../../DATA/featured_telco.csv")

How should the data be prepared for model training?

Machine Learning models require separating the input features from the target variable before training.

In [6]:
X = df.drop(columns=["Churn Label"])

y = df["Churn Label"]

In [7]:
print(X.shape)
print(y.shape)

(7032, 31)
(7032,)


How should the dataset be split into training and testing sets?

The model should learn from one part of the data and be tested on unseen data.

80% of the data will be used for training and 20% for testing.

In [8]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [9]:
print(X_train.shape)
print(X_test.shape)

(5625, 31)
(1407, 31)


Is the training dataset balanced?

An imbalanced dataset may cause the model to favor the majority class.

The training dataset is imbalanced because non-churn customers are more frequent than churn customers.

In [10]:
y_train.value_counts()

Churn Label
0    4130
1    1495
Name: count, dtype: int64

Should oversampling be applied?

Oversampling increases the number of minority class samples to improve model learning.

Both classes now contain the same number of observations

In [11]:
smote = SMOTE(random_state=42)

X_train_smote, y_train_smote = smote.fit_resample(
    X_train,
    y_train
)

In [12]:
print(y_train_smote.value_counts())

Churn Label
0    4130
1    4130
Name: count, dtype: int64


Which machine learning models will be trained?



In [13]:
models = {
    "LogisticRegression": LogisticRegression(max_iter=5000, random_state=42),

    "Decision Tree": DecisionTreeClassifier(random_state=42),

    "Random Forest": RandomForestClassifier(random_state=42),

    "Gradient Boosting": GradientBoostingClassifier(random_state=42)
}

In [14]:
models

{'LogisticRegression': LogisticRegression(max_iter=5000, random_state=42),
 'Decision Tree': DecisionTreeClassifier(random_state=42),
 'Random Forest': RandomForestClassifier(random_state=42),
 'Gradient Boosting': GradientBoostingClassifier(random_state=42)}

How can the selected models be trained?

Each model should learn the relationship between customer features and churn.

In [15]:
trained_models = {}

for name, model in models.items():

    model.fit(X_train_smote, y_train_smote)

    trained_models[name] = model

    print(name, "trained successfully.")

LogisticRegression trained successfully.
Decision Tree trained successfully.
Random Forest trained successfully.
Gradient Boosting trained successfully.


All selected models were successfully trained using the balanced training dataset.

Can cross-validation improve model reliability?

Cross-validation evaluates the model on multiple training subsets instead of only one split.

Cross-validation provides a more reliable estimate of the model's performance across different subsets of

In [16]:
for name, model in models.items():

    scores = cross_val_score(
        model,
        X_train_smote,
        y_train_smote,
        cv=5,
        scoring="accuracy"
    )

    print(name)

    print(scores)

    print("Mean Accuracy:", scores.mean())

    print("-"*40)

LogisticRegression
[0.75121065 0.78874092 0.8716707  0.86440678 0.88438257]
Mean Accuracy: 0.8320823244552059
----------------------------------------
Decision Tree
[0.71489104 0.73547215 0.86440678 0.84806295 0.86198547]
Mean Accuracy: 0.8049636803874092
----------------------------------------
Random Forest
[0.77360775 0.8153753  0.9098063  0.90677966 0.91101695]
Mean Accuracy: 0.863317191283293
----------------------------------------
Gradient Boosting
[0.7409201  0.78934625 0.90012107 0.88559322 0.90556901]
Mean Accuracy: 0.8443099273607748
----------------------------------------


In [17]:
print(X.columns.tolist())

['Gender', 'Senior Citizen', 'Partner', 'Dependents', 'Tenure Months', 'Phone Service', 'Multiple Lines', 'Online Security', 'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV', 'Streaming Movies', 'Paperless Billing', 'Monthly Charges', 'Total Charges', 'CLTV', 'Contract_Month-to-month', 'Contract_One year', 'Contract_Two year', 'Payment Method_Bank transfer (automatic)', 'Payment Method_Credit card (automatic)', 'Payment Method_Electronic check', 'Payment Method_Mailed check', 'Internet Service_DSL', 'Internet Service_Fiber optic', 'Internet Service_No', 'Average Charges', 'CLTV per Month', 'High Monthly Charge', 'Loyal Customer']


Which data will be used in the next stage?

Logistic Regression → 83%

Decision Tree → 80.6%

Random Forest → 86.4% (highest accuarcy)

Gradient Boosting → 84.5%




إحنا بنعمل SMOTE على X_train و y_train فقط، مش على الداتا كلها.

لأن لو عملنا SMOTE قبل الـ Train/Test Split يبقى الموديل شاف معلومات من الـ Test Data، وده يعتبر Data Leakage ويخلي نتائج التقييم غير حقيقية.

## Conclusion

The dataset was successfully divided into training and testing sets.

SMOTE was applied to balance the training data.

Four machine learning models were trained using the balanced dataset.

Cross-validation was performed to estimate each model's generalization capability.

The trained models are now ready for evaluation in the next notebook.